src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+                    if (signatureProvider == null)
+                        throw LogHelper.LogExceptionMessage(new SignedHttpRequestInvalidSignatureException(LogHelper.FormatInvariant(LogMessages.IDX23000, popKey?.ToString() ?? "Null", jwtSignedHttpRequest.Alg)));
+
+                    try
+                    {
+                        var encodedBytes = Encoding.UTF8.GetBytes(jwtSignedHttpRequest.EncodedHeader + "." + jwtSignedHttpRequest.EncodedPayload);
+                        var signature = Base64UrlEncoder.DecodeBytes(jwtSignedHttpRequest.EncodedSignature);
+
+                        if (signatureProvider.Verify(encodedBytes, signature))
+                            return popKey;
+                    }
+                    finally
+                    {
+                        popKey.CryptoProviderFactory.ReleaseSignatureProvider(signatureProvider);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    exceptionStrings.AppendLine(ex.ToString());
+                }
+            }
+
+            throw LogHelper.LogExceptionMessage(new SignedHttpRequestInvalidSignatureException(LogHelper.FormatInvariant(LogMessages.IDX23009, exceptionStrings.ToString())));
+        }
+
+        /// <summary>
+        /// Validates the signed http request lifetime ("ts").
+        /// </summary>
+        /// <param name="signedHttpRequest">A SignedHttpRequest.</param>
+        /// <param name="signedHttpRequestValidationContext">A structure that wraps parameters needed for SignedHttpRequest validation.</param>
+        /// <remarks>
+        /// This method will be executed only if <see cref="SignedHttpRequestValidationParameters.ValidateTs"/> is set to <c>true</c>.
+        /// </remarks>    
+        protected virtual void ValidateTsClaim(SecurityToken signedHttpRequest, SignedHttpRequestValidationContext signedHttpRequestValidationContext)
+        {
+            if (signedHttpRequest == null)
+                throw LogHelper.LogArgumentNullException(nameof(signedHttpRequest));
+
+            if (!(signedHttpRequest is JsonWebToken jwtSignedHttpRequest))
+                throw LogHelper.LogExceptionMessage(new SignedHttpRequestValidationException(LogHelper.FormatInvariant(LogMessages.IDX23031, signedHttpRequest.GetType(), typeof(JsonWebToken), signedHttpRequest)));
+
+            if (!jwtSignedHttpRequest.TryGetPayloadValue(SignedHttpRequestClaimTypes.Ts, out long tsClaimValue))
+                throw LogHelper.LogExceptionMessage(new SignedHttpRequestInvalidTsClaimException(LogHelper.FormatInvariant(LogMessages.IDX23003, SignedHttpRequestClaimTypes.Ts)));
+
+            DateTime utcNow = DateTime.UtcNow;
+            DateTime signedHttpRequestCreationTime = EpochTime.DateTime(tsClaimValue);
+            DateTime signedHttpRequestExpirationTime = signedHttpRequestCreationTime.Add(signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.SignedHttpRequestLifetime);
```

Comment:
Usually JWT validation libraries have the ClockSkew setting. This setting is not intended for the signed http request?

Reply:
SignedHttpRequestLifetime is set to 5 minutes (by default), but it's configurable via SignedHttpRequestValidationParameters.
For signed http request, the clock skew is moved to the client side i.e. clients should take care of computing the skew and adjusting the time.

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/7d9235b0dd4aea3b6082c4c1080408ae377f4b06/src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestCreationParameters.cs#L98-L102

Reply:
@GeoK and @brentschmaltz and may I ask why the change in the skewing logic compared to regular JWT tokens?

Considerations is that the actual skew could be both + and - and the TimeAdjustment is always in just one direction.